atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssigner.java

```diff
@@ -0,0 +1,216 @@
+        metrics.bucketStateForIteration(state);
+        return state;
+    }
+
+    private final class BucketStateVisitor implements BucketAssignerState.Visitor<BucketState> {
+        private final long bucketIdentifier;
+
+        private BucketStateVisitor(long bucketIdentifier) {
+            this.bucketIdentifier = bucketIdentifier;
+        }
+
+        @Override
+        public BucketState visit(BucketAssignerState.Start start) {
+            metrics.progressedToBucketIdentifier(bucketIdentifier);
+            metrics.bucketAssignerState(start);
+            long openWallClockTimeMillis = puncherStore.getMillisForTimestamp(start.startTimestampInclusive());
+            long closeWallClockTimeMillis = getWallClockCloseTimeGivenOpenTime(openWallClockTimeMillis);
+            if (Instant.now(clock).toEpochMilli() < closeWallClockTimeMillis) {
+                return tryTransitionAndExecuteNextState(
+                        start, BucketAssignerState.opening(start.startTimestampInclusive()));
+            } else {
+                long endTimestampExclusive = puncherStore.get(closeWallClockTimeMillis);
+                return tryTransitionAndExecuteNextState(
+                        start,
+                        BucketAssignerState.immediatelyClosing(start.startTimestampInclusive(), endTimestampExclusive));
+            }
+        }
+
+        @Override
+        public BucketState visit(BucketAssignerState.Opening open) {
+            metrics.bucketAssignerState(open);
+            WriteState successfullyWroteAllBuckets = bucketWriter.writeToAllBuckets(
+                    bucketIdentifier, Optional.empty(), TimestampRange.of(open.startTimestampInclusive(), -1));
+            if (successfullyWroteAllBuckets == WriteState.FAILURE) {
+                return BucketState.CONTENTION_ON_WRITES;
+            }
+
+            BucketAssignerState waitingUntilCloseable =
+                    BucketAssignerState.waitingUntilCloseable(open.startTimestampInclusive());
+            sweepBucketAssignerStateMachineTable.updateStateMachineForBucketAssigner(
+                    BucketStateAndIdentifier.of(bucketIdentifier, open),
+                    BucketStateAndIdentifier.of(bucketIdentifier, waitingUntilCloseable));
+            return BucketState.INCOMPLETE;
+        }
+
+        @Override
+        public BucketState visit(BucketAssignerState.WaitingUntilCloseable waitingClose) {
+            metrics.bucketAssignerState(waitingClose);
+            long openWallClockTimeMillis = puncherStore.getMillisForTimestamp(waitingClose.startTimestampInclusive());
+            long closeWallClockTimeMillis = getWallClockCloseTimeGivenOpenTime(openWallClockTimeMillis);
+
+            if (Instant.now(clock).toEpochMilli() < closeWallClockTimeMillis) {
+                return BucketState.INCOMPLETE;
+            }
+
+            long finalTimestamp = puncherStore.get(closeWallClockTimeMillis);
```

Comment:
Do we have protection here/elsewhere against clock drift situations which could mean that the final timestamp is less than the start timestamp?

Reply:
Documenting what was decided: We'll grab the latest timestamp from timelock, then take min(start + X, latestTimestamp) as the closing time.